### ISSUE TITLE
`cfg_not_test`

### ISSUE THREAD

**Issue by JonathanWoollett-Light** [rust-lang/rust-clippy#11234]
### What it does

Disallows usage of any conditional compilation that always excludes code from test builds.

### Advantage

- Prevents code being excluded from tests builds. Preventing the case where a codebase appears to have 100% coverage while having untested code.
- Enforces simplicity in testing, guarding against excessive mocking anti-pattern.

### Drawbacks

- May prevent rare legitimate use cases for excluding code from test runs.

### Example

The following would trigger this lint:

```rust
#[cfg(not(test))]
```

```rust
#[cfg(all(debug_assertions, not(test)))]
```

```rust
#[cfg(not(any(not(debug_assertions), test)))]
```

**Comment by blyxyas** [comment]
I feel like the drawback is too heavy for an allow-by-default lint. What do you think about `restriction`?

**Comment by JonathanWoollett-Light** [comment]
@blyxyas I would agree it makes sense for it to be a `restriction` lint.
